src/panic.md

```diff
@@ -0,0 +1,98 @@
+# Panic
+
+Rust provides the ability to "panic" upon encountering an error that is not
+part of a function's signature; such an error is typically not expected to be
+recoverable within the context in which the error is encountered.
+
+Some language constructs, such as out-of-bounds [array indexing], panic
+automatically. There are also language features that provide a level of control
+over panic behavior:
+* A [_panic runtime_](#panic-runtimes) defined how a panic is handled during
+  runtime.
+* [FFI ABIs](items/functions.md#unwinding) may alter how panics behave.
+
+> **Note**: The Standard Library provides the capability to explicitly panic
+> via the [`panic!` macro][macro-panic].
+
+## Unwinding
+
+Panicking may either be recoverable or non-recoverable, though it can be
+configured (via `panic=abort`) to always be non-recoverable. (The converse is
+not true: `panic=unwind` does not guarantee that all panics are recoverable,
+only that panicking via the `panic!` macro and similar standard library
+mechanisms is recoverable.) A recoverable panic "unwinds" Rust frames, just as
+C++'s `throw` unwinds C++ frames. This means that as the panic traverses Rust
+frames, live objects in those frames that [implement `Drop`][destructors] will
+have their `drop` methods called. Thus, if panic recovery does occur (for
+instance at a thread boundary), the objects will have been "cleaned up" just as
+if they had gone out of scope normally.
```

Comment:
No, because that issue only applies if panic recovery _does not_ occur, i.e. the process aborts.

I suppose one could argue that panic recovery "starts to" occur and then is "interrupted," but the whole problem in that issue, as I understand it, is that this isn't really how most unwinding implementations think about this kind of early termination: if you're going to abort the process anyway in the process of "recovering," then you haven't really "recovered" at any point.

Reply:
Right.  Probably I mean this part specifically though:

> This means that as the panic traverses Rust frames, live objects in those frames that implement `Drop` will have their `drop` methods called.

This language isn't conditioned in the text on panic recovery occurring (the language later is, but that language is stated as a consequence of this language).

Or more in context:

> Panicking may either be recoverable or non-recoverable, though it can be configured (via `panic=abort`) to always be non-recoverable. (The converse is not true: `panic=unwind` does not guarantee that all panics are recoverable, only that panicking via the `panic!` macro and similar standard library mechanisms is recoverable.) A recoverable panic "unwinds" Rust frames, just as C++'s `throw` unwinds C++ frames. This means that as the panic traverses Rust frames, live objects in those frames that implement `Drop` will have their `drop` methods called.

So a reasonable reader could conclude that if `panic=unwind` is set, then a panic due to the `panic!` macro is recoverable, and all recoverable panics unwind Rust frames, and that live objects in those frames are therefore guaranteed to be dropped.

Reply:
Okay, I think this is better now; it emphasizes the recovery more, and correctly identifies `thread::spawn` (rather than `JoinHandle::join`) as the mechanism of recovery.